[FEATURE][3] Add Favorites

mathiramilo-lit opened the pull request:
# Add Favorites - 3

## Description ⭐

I've implemented add to favorites storing favorite products in local storage.
I've also rebase and merge the URL state handling PR.

### Type of change 🏃

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

## Screenshots 📷

mathiramilo-lit committed:
feat: add favorites

mathiramilo-lit committed:
feat: add url state handling

danielaWaldeck95 commented on src/App.tsx:
```diff
@@ -35,6 +55,21 @@ function App() {
+  useEffect(() => {
+    const newParams = new URLSearchParams();
+
+    if (query) newParams.set('search', query);
+    if (sort.sortBy) newParams.set('sortBy', sort.sortBy);
+    if (sort.order) newParams.set('order', sort.order);
```
```
    query && newParams.set('search', query);
    sort.sortBy && newParams.set('sortBy', sort.sortBy);
    sort.order && newParams.set('order', sort.order);
```

mathiramilo-lit replied:
```
Expected an assignment or function call and instead saw an expression
.eslint[@typescript-eslint/no-unused-expressions](https://typescript-eslint.io/rules/no-unused-expressions)
```

I'm encountering the error "Expected an assignment or function call and instead saw an expression" from ESLint. What could be causing this?

danielaWaldeck95 replied:
Sorry ESLint is right, mybad, the && conditional is being flagged as an unused expression.

Just keep the ifs. I would add the {} anyways, it adds a few more codelines but prevents errors and improves readability

mathiramilo-lit replied:
Perfect. So I'll keep the if statements and add the {} for improved readability and error prevention. Thanks for the help!

danielaWaldeck95 commented on src/components/ui/loader.tsx:
```diff
@@ -0,0 +1,16 @@
+  return (
+    <div
+      className={cn(
+        'h-4 w-4 animate-spin rounded-full border-t-2 border-t-white',
```
https://tailwindcss.com/docs/size you can replace h-4 w-4 by size-4

mathiramilo-lit replied:
Thanks for the tip! I didn’t know about that Tailwind shortcut.

danielaWaldeck95 commented on src/components/ui/sort-by.tsx:
```diff
@@ -1,17 +1,39 @@
+      : 'Sort By',
+  );
+
+  useEffect(() => {
+    if (!showFavorites && !sort.sortBy && !sort.order) {
+      setLabel('Sort By');
+    }
+  }, [showFavorites, sort]);
```
I think this useEffect is not needed, the component will re-render if showFavorites and sort states change on the parent

mathiramilo-lit replied:
The purpose of this useEffect is to execute when the showFavorites state changes in the App component. If I remove it, the functionality stops working. I haven’t been able to figure out a cleaner approach. Do you have any suggestions?

danielaWaldeck95 commented on src/hooks/use-products.ts:
```diff
@@ -29,7 +30,7 @@ export const useProducts = ({ query, sort }: UseProductsOptions) => {
     try {
       const res = await axios.get(
-        `https://dummyjson.com/products${query ? `/search?q=${query}&` : '?'}limit=${LIMIT}&skip=${pagination.skip}${sort ? `&sortBy=${sort.sortBy}&order=${sort.order}` : ''}`,
+        `https://dummyjson.com/products${query ? `/search?q=${query}&` : '?'}limit=${LIMIT}&skip=${pagination.skip}${sort?.sortBy && sort.order ? `&sortBy=${sort.sortBy}&order=${sort.order}` : ''}`,
```
same suggestion H left on the other pr

mathiramilo-lit replied:
Great suggestion! I'm fixing that.

danielaWaldeck95 commented on src/hooks/use-products.ts:
```diff
@@ -50,9 +51,8 @@ export const useProducts = ({ query, sort }: UseProductsOptions) => {
       try {
         const res = await axios.get(
-          `https://dummyjson.com/products${query ? `/search?q=${query}&` : '?'}limit=${LIMIT}${sort ? `&sortBy=${sort.sortBy}&order=${sort.order}` : ''}`,
+          `https://dummyjson.com/products${query ? `/search?q=${query}&` : '?'}limit=${LIMIT}${sort?.sortBy && sort.order ? `&sortBy=${sort.sortBy}&order=${sort.order}` : ''}`,
```
same suggestion H left on the other pr

danielaWaldeck95 commented on src/hooks/use-debounce.ts:
```diff
@@ -1,17 +1,21 @@
 
   useEffect(() => {
-    const handler = setTimeout(() => {
+    const debounce = setTimeout(() => {
```
rename to timeoutId

danielaWaldeck95 reviewed (commented):
Left some comments

jsarrolt commented on src/components/products/product-card.tsx:
```diff
@@ -1,14 +1,19 @@
-import { Heart } from '../assets';
-import { Product } from '../types';
+import { Product } from '../../types';
+import { HeartFill, HeartOutline } from '../../assets';
+import { useFavorites } from '../../context';
+import { useMemo } from 'react';
```
Consider setting up [typescript barrel generator](https://marketplace.visualstudio.com/items?itemName=eliostruyf.vscode-typescript-exportallmodules) and [path aliases](https://blog.logrocket.com/using-path-aliases-cleaner-react-typescript-imports/) for the project

mathiramilo-lit replied:
Thanks for the suggestion! I went ahead and added the path aliases

jsarrolt commented on src/components/products/product-card.tsx:
```diff
@@ -1,14 +1,19 @@
+  const { favorites, addFavorite, removeFavorite } = useFavorites();
+
+  const isFavorite = useMemo(
+    () => Boolean(favorites.find((p) => p.id === product.id)),
```
```suggestion
    () => favorites.some((p) => p.id === product.id),
```

jsarrolt commented on src/components/ui/sort-by.tsx:
```diff
@@ -1,17 +1,39 @@
   options: SortOption[];
-  setSort: (sortBy: ProductField, order: Order) => void;
+  sort: SortState;
+  setSort: (sortBy: SortByType | undefined, order: Order | undefined) => void;
```
```suggestion
setSort: (sortBy?: SortByType, order?: Order) => void;
```

jsarrolt commented on src/App.tsx:
```diff
@@ -11,17 +11,37 @@ import {
 
-export type SortState = Omit<SortOption, 'label'>;
+export type SortState = {
+  sortBy: SortByType | undefined;
+  order: Order | undefined;
```
```suggestion
  sortBy?: SortByType;
  order?: Order;
```

jsarrolt reviewed (commented):
Great work!

mathiramilo-lit committed:
chore: add import alias

Update tsconfig files and vite.config.ts to add import alias, and create cspell.json file to add unrecognized words.

mathiramilo-lit committed:
fix: pr comments

* update axios calls to use params
* change "debounce" to "timeoutId" in useDebounce
* change SortBy component: rename to "SortDropdown", object-parameter to setSort and remove useEffect
* add ? for possible undefined values
